### ISSUE TITLE
Find atropos for a block

### ISSUE THREAD

**Issue by grasphoper** [Fantom-foundation/lachesis-dag-tool#5]
Hi @rus-alex 

I'm trying to understand how fantom protocol consensus works more in-depth. I've stumbled upon this repo and I can see that you're arriving at block atropos by doing [this](https://github.com/Fantom-foundation/lachesis-dag-tool/blob/418f4060cf2d8325b45f9aa7a759c07e12bb1f5f/dagreader/reader.go#L150):
`atropos := hash.Event(blk.Hash())`

Doing this yields `event 0x... not found` for any block I try it with (I'm using `ftmclient.Client` and the `.GetEvent` method). Do you have any hint as to why this might work? How do I find block atropos without modifying the `go-opera` code?

Thanks in advance!

**Comment by grasphoper** [comment]
I found this line of code in `go-opera`:

`Hash:       common.Hash(block.Atropos),` 
from function `func ToEvmHeader(block *inter.Block, index idx.Block, prevHash hash.Event, rules opera.Rules) *EvmHeader {` , which confirms that Atropos == block.Hash()

This brings me to the next question. Do I need to be running an archive node for this to work?

**Comment by rus-alex** [comment]
Hi @sockerdev !
The lachesis-dag-tool/dagreader is not about consensus. It just reads opera DAG events from [node's](https://github.com/Fantom-foundation/go-opera) API. So yes, the node should be runned and synced with {main,test,other}net.

**Comment by grasphoper** [comment]
@rus-alex 
Let me rephrase that. I would like to get an "eventBlock representation" of a single block by its number. So, a "dag representation" of one single block

I am running a fully synced mainnet API node, and have `dag,abft` namespaces open. The method `dag_getEvent` is available, but this code panics with `event 0x490c4b4ac616ebc69191df66241f21399a2277238c007a183512ae229f06141a not found`:

```
blockNumber := big.NewInt(52370764)
b, err := c.BlockByNumber(context.Background(), blockNumber)
if err != nil {
panic(err)
}
atropos := hash.Event(b.Hash())
ev, err := c.GetEvent(context.Background(), atropos)
if err != nil {
panic(err)
}
fmt.Println(ev.Parents())
```

This doesn't only happen with block `52370764`, but with any block

**Comment by rus-alex** [comment]
@sockerdev , it looks like a problem with your fully synced mainnet API node.
I run your code example on public mainnet API ( https://rpcapi.fantom.network/ ) and got the event well.

Please note that block `52370764` hash is `0x0002b6450000005ca7fcd13e9f0e08d10f5e8f7546cd644def74af9b215a2e04` (see https://ftmscan.com/block/52370764), I have no idea why it is `0x490c4b4ac616ebc69191df66241f21399a2277238c007a183512ae229f06141a` in your example.

**Comment by grasphoper** [comment]
@rus-alex I didn't investigate too deeply yet, but I think I know what the issue is.

It's that `ethclient.Client` (from [go-ethereum repo](github.com/ethereum/go-ethereum/ethclient) `v1.10.26` doesn't provide correct response for the method `BlockByNumber`.

Both my node and public RPC you provided, `https://rpcapi.fantom.network/` return `0x490c4b4ac616ebc69191df66241f21399a2277238c007a183512ae229f06141a` as hash for block `52370764`, but if I instead start using lower-level calls, the problem goes away. This code returns valid hash:

```
var r map[string]interface{}
e := c.rpcClient.CallContext(ctx, &r, "eth_getBlockByNumber", hexutil.EncodeBig(number), false)
```

So, that's an issue specific to the `go-ethereum`. Although it does affect your code as well. (for example in `ftmclient.Client`).

Anyway, thank you for the help and cooperation!
